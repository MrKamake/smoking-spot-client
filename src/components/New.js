import React, { useState } from 'react';
import './New.scss';
import { Link } from 'react-router-dom';

import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng
} from 'react-places-autocomplete';
import jwt from 'jsonwebtoken';

let userToken = localStorage.getItem('jwtToken');
let decodedToken;
if (userToken) {
  decodedToken = jwt.verify(userToken, process.env.REACT_APP_SECRET_KEY);
}

const New = ({ onSubmit, failMessage }) => {
  const [imagePreviewUrl, setImagePreviewUrl] = useState('');
  const [file, setFile] = useState(null);
  const [title, setTitle] = useState('');
  const [address, setAddress] = useState('');
  const [detailAddress, setDetailAddress] = useState('');
  const [location, setLocation] = useState({});

  const _handleSelect = addressData => {
    geocodeByAddress(addressData)
      .then(result => getLatLng(result[0]))
      .then(latLng => setLocation(latLng))
      .catch(err => console.log('error', err));
  };

  const _handlePlusPhoto = ev => {
    ev.preventDefault();

    const reader = new FileReader();
    const file = ev.target.files[0];

    reader.onloadend = () => {
      setFile(file);
      setImagePreviewUrl(reader.result);
    };
    reader.readAsDataURL(file);
  };

  const _handleSubmit = ev => {
    ev.preventDefault();

    const imageData = new FormData();
    imageData.append('file', file);

    const spotInfo = {
      email: decodedToken.email,
      title,
      address: `${address + detailAddress}`,
      location: { type: 'Point', coordinates: [location.lng, location.lat] }
    };

    onSubmit(imageData, spotInfo);
  };

  return (
    <>
      {failMessage && alert(failMessage)}
      <Link to='/home' className='sub-logo'>
        <div className='sub-logo-smoking'>SMOKING</div>
        <div className='sub-logo-spot'>SPOT</div>
      </Link>
      <form
        className='new-formbox'
        onSubmit={_handleSubmit}
        encType='multipart/form-data'
      >
        <div className='new-header-box'>
          <div className='new-text'>New</div>
          <div className='photo-plus-box'>
            <input
              id='form-file'
              className='photo-description'
              type='file'
              name='file'
              onChange={_handlePlusPhoto}
            />
            <label htmlFor='form-file' className='photo-plus-button'>
              +photo
            </label>
          </div>
        </div>
        <div className='input-text-group'>
          <div className='previe-image'>
            {imagePreviewUrl ? (
              <img className={`new-img`} alt='img' src={imagePreviewUrl} />
            ) : (
              <div className={`new-img-${!!imagePreviewUrl}`}></div>
            )}
          </div>
          <input
            className='input-title group'
            type='text'
            name='title'
            onChange={ev => setTitle(ev.target.value)}
            placeholder='title'
          />
          <PlacesAutocomplete
            value={address}
            onChange={addressData => setAddress(addressData)}
            onSelect={_handleSelect}
          >
            {({
              getInputProps,
              suggestions,
              getSuggestionItemProps,
              loading
            }) => (
              <>
                <input
                  {...getInputProps({
                    placeholder: 'Search Places ...',
                    className: 'location-search-input group'
                  })}
                />
                <div className='autocomplete-dropdown-container'>
                  {loading && <div>Loading...</div>}
                  {suggestions.map(suggestion => {
                    return (
                      <div
                        {...getSuggestionItemProps(suggestion, {
                          className: suggestion.active
                            ? 'suggestion-item--active'
                            : 'suggestion-item',
                          style: suggestion.active
                            ? { backgroundColor: '#fafafa', cursor: 'pointer' }
                            : { backgroundColor: '#ffffff', cursor: 'pointer' }
                        })}
                      >
                        <span>{suggestion.description}</span>
                      </div>
                    );
                  })}
                </div>
              </>
            )}
          </PlacesAutocomplete>
          <input
            className='group'
            type='text'
            name='detail-address'
            onChange={ev => setDetailAddress(ev.target.value)}
            placeholder='detail-address'
          />
          <input className='submit-box group' name='submit' type='submit' value='Submit' />
        </div>
      </form>
    </>
  );
};

export default New;
