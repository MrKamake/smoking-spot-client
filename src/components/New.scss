.new-formbox {
  text-align: center;
  .new-header-box {
    position: relative;
    width: 100%;
    height: 30px;
    padding-bottom: 4px;
    .new-text {
      position: absolute;
      left: 4%;
      font-size: 1.8rem;
      font-weight: bold;
      color: dodgerblue;
    }
    .photo-plus-box {
      position: absolute;
      right: 1.5%;
      height: 20px;
      padding: 10px 0;
      .photo-description {
        display: none;
      }
      .photo-plus-button {
        padding: 5px 10px;
        border-radius: 4px;
        color: white;
        background-color: darkgray;
      }
    }
  }
  .input-text-group {
    width: 100%;
    .previe-image {
      margin: 15px 0;
      .new-img {
        display: inline-block;
        width: 100%;
      }
      .new-img-false {
        display: inline-block;
        width: 365px;
        height: 365px;
        color: lightgrey;
        background: lightgrey;
      }
    }
    .group {
      display: inline-block;
      margin: 15px 0;
      width: 350px;
      height: 40px;
      line-height: 50px;
      padding: 1px 10px;
      border: 1px solid lightgrey;
      text-decoration: none;
      font-size: 1rem;
      border-radius: 5px;
    }
    .submit-box {
      bottom: 0;
      border: 0;
      font-size: 1.2rem;
      font-weight: bold;
      color: white;
      background: rgb(80, 170, 121);
    }
  }
}
