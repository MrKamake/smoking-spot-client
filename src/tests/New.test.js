import React from 'react';
import { shallow } from 'enzyme';
import New from '../components/New';

describe('<New />', () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation(init => [init, setState]);

  beforeEach(() => {
    wrapper = shallow(<New />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const photoButton = wrapper.find('.photo-plus-button');

    expect(photoButton.contains('+photo')).toBe(true);
    expect(wrapper.contains(<div className='new-text'>New</div>)).toBe(true);
    expect(
      wrapper.contains(<div className='sub-logo-smoking'>SMOKING</div>)
    ).toBe(true);
    expect(wrapper.contains(<div className='sub-logo-spot'>SPOT</div>)).toBe(
      true
    );
  });

  // it('renders changed states on input-text typing', () => {
  //   const onSubmit = jest.fn();
  //   wrapper = shallow(<New onSubmit={onSubmit} />);
  //   let titleInput = wrapper.find('input[name="title"]');
  //   const submitButton = wrapper.find('input[name="submit"]');

  //   titleInput.simulate('change', {
  //     target: {
  //       value: 'Title-test'
  //     }
  //   });

  //   submitButton.simulate('click');
  //   titleInput = {
  //     target: {
  //       value: 'Title-test'
  //     }
  //   };
  //   expect(onSubmit).toHaveBeenCalledWith(titleInput);
  // });
});
