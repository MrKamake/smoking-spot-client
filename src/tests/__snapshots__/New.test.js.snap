// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<New /> matches snapshot 1`] = `
<Fragment>
  <Link
    className="sub-logo"
    to="/home"
  >
    <div
      className="sub-logo-smoking"
    >
      SMOKING
    </div>
    <div
      className="sub-logo-spot"
    >
      SPOT
    </div>
  </Link>
  <form
    className="new-formbox"
    encType="multipart/form-data"
    onSubmit={[Function]}
  >
    <div
      className="new-header-box"
    >
      <div
        className="new-text"
      >
        New
      </div>
      <div
        className="photo-plus-box"
      >
        <input
          className="photo-description"
          id="form-file"
          name="file"
          onChange={[Function]}
          type="file"
        />
        <label
          className="photo-plus-button"
          htmlFor="form-file"
        >
          +photo
        </label>
      </div>
    </div>
    <div
      className="input-text-group"
    >
      <div
        className="previe-image"
      >
        <div
          className="new-img-false"
        />
      </div>
      <input
        className="input-title group"
        name="title"
        onChange={[Function]}
        placeholder="title"
        type="text"
      />
      <PlacesAutocomplete
        debounce={200}
        highlightFirstSuggestion={false}
        onChange={[Function]}
        onError={[Function]}
        onSelect={[Function]}
        searchOptions={Object {}}
        shouldFetchSuggestions={true}
        value=""
      >
        <Component />
      </PlacesAutocomplete>
      <input
        className="group"
        name="detail-address"
        onChange={[Function]}
        placeholder="detail-address"
        type="text"
      />
      <input
        className="submit-box group"
        name="submit"
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</Fragment>
`;
